core/vm/access_list_tracer.go

```diff
@@ -0,0 +1,171 @@
+	idx, addrPresent := al.addresses[address]
+	if !addrPresent || idx == -1 {
+		delete(al.addresses, address)
+	}
+}
+
+// copy creates an independent copy of an accessList.
+func (a *accessList) copy() *accessList {
+	cp := newAccessList()
+	for k, v := range a.addresses {
+		cp.addresses[k] = v
+	}
+	cp.slots = make([]map[common.Hash]struct{}, len(a.slots))
+	for i, slotMap := range a.slots {
+		newSlotmap := make(map[common.Hash]struct{}, len(slotMap))
+		for k := range slotMap {
+			newSlotmap[k] = struct{}{}
+		}
+		cp.slots[i] = newSlotmap
+	}
+	return cp
+}
+
+// toAccesslist converts the accesslist to a types.AccessList.
+// The accessList internally uses mappings which makes it
+// easier to add and remove values.
+func (a *accessList) toAccessList() types.AccessList {
+	acl := make(types.AccessList, 0, len(a.addresses))
+	for addr, idx := range a.addresses {
+		var tuple types.AccessTuple
+		tuple.Address = addr
+		// addresses without slots are saved as -1
+		if idx >= 0 {
+			keys := make([]common.Hash, 0, len(a.slots[idx]))
+			for key := range a.slots[idx] {
+				keys = append(keys, key)
+			}
+			tuple.StorageKeys = keys
+		}
+		acl = append(acl, tuple)
+	}
+	return acl
+}
+
+type accessListTracer struct {
+	list *accessList
+}
+
+// NewAccessListTracer creates a new tracer that can generate AccessLists.
+// An optional AccessList can be specified to occupy slots and addresses in
+// the resulting accesslist.
+func NewAccessListTracer(acl types.AccessList) *accessListTracer {
+	list := newAccessList()
+	for _, al := range acl {
+		list.addAddress(al.Address)
+		for _, slot := range al.StorageKeys {
+			list.addSlot(al.Address, slot)
+		}
+	}
+	return &accessListTracer{
+		list: list,
+	}
+}
+
+func (a *accessListTracer) CaptureStart(env *EVM, from common.Address, to common.Address, create bool, input []byte, gas uint64, value *big.Int) {
+}
+
+func (a *accessListTracer) CaptureState(env *EVM, pc uint64, op OpCode, gas, cost uint64, scope *ScopeContext, rData []byte, depth int, err error) {
+	stack := scope.Stack
+	if (op == SLOAD || op == SSTORE) && stack.len() >= 1 {
+		slot := common.Hash(stack.data[stack.len()-1].Bytes32())
+		a.list.addSlot(scope.Contract.Address(), slot)
+	}
+	if (op == EXTCODECOPY || op == EXTCODEHASH || op == EXTCODESIZE || op == BALANCE || op == SELFDESTRUCT) && stack.len() >= 1 {
+		address := common.Address(stack.data[stack.len()-1].Bytes20())
+		a.list.addAddress(address)
+	}
+	if (op == DELEGATECALL || op == CALL || op == STATICCALL || op == CALLCODE) && stack.len() >= 5 {
+		address := common.Address(stack.data[stack.len()-2].Bytes20())
+		a.list.addAddress(address)
+	}
+}
+
+func (*accessListTracer) CaptureFault(env *EVM, pc uint64, op OpCode, gas, cost uint64, scope *ScopeContext, depth int, err error) {
+}
+
+func (*accessListTracer) CaptureEnd(output []byte, gasUsed uint64, t time.Duration, err error) {}
+
+// GetAccessList returns the current AccessList maintained by the tracer.
```

Comment:
Go code rarely calls something `GetXyz`, since `Xyz` often suffices. Here calling it `AccessList` and below `UnpreparedAccessList` is meaningful enough to not require an extra `Get` imho.

Reply:
Also, perhaps `UnpreparedAccessList` is a bit of a strange name, what is `unprepared` about it?

I'd suggest calling `UnpreparedAccessList` simply `AccessList` and this original `AccessList` -> `RawAccessList` since this specific one is the more special and the other is the final. Or I'd accept prefixing the Unprepared one too, but do we have a slightly more meaningful name? :D

Reply:
fixed, `GetAccessList()` -> `RawAccessList` and `GetUnpreparedAccessList` -> `AccessList`.
The "unprepare" was in response to the prepare step used in `core/state/access_list.go`